docs/features/async-streams.md

```diff
@@ -164,3 +170,135 @@ Similarly, the kick-off method is much like those of regular iterator methods:
     return result;
 }
 ```
+
+#### Disposal
+
+Iterator and async-iterator methods need disposal because their execution steps are controlled by the caller, which could choose to dispose the enumerator before getting all of its elements.
+For example, `foreach (...) { if (...) break; }`.
+In contrast, async methods continue running autonomously until they are done. They are never left suspended in the middle of execution from the caller's perspective, so they don't need to be disposed.
+
+In summary, disposal of an async-iterator works based on four design elements:
+- `yield return` (jumps to finally when resuming in dispose mode)
+- `yield break` (enters dispose mode and jumpps to finally)
+- `finally` (after a `finally` we jump to the next one)
+- `DisposeAsync` (enters dispose mode and resumes execution)
+
+The caller of an async-iterator method should only call `DisposeAsync()` when the method completed or was suspended by a `yield return`.
+`DisposeAsync` sets a flag on the state machine ("dispose mode") and (if the method wasn't completed) resumes the execution from the current state.
+The state machine can resume execution from a given state (even those located within a `try`).
+When the execution is resumed in dispose mode, it jump straight to the relevant `finally`.
+`finally` blocks may involve pauses and resumes, but only for `await` expressions. As a result of the restrictions imposed on `yield return` (described above), dispose mode never runs into that statement.
```

Comment:
>that statement [](start = 189, length = 14)

What is `that statement`?

Reply:
`yield return`
I'll make it explicit

---
In reply to: [240414615](https://github.com/dotnet/roslyn/pull/31527#discussion_r240414615) [](ancestors = 240414615)